src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/InstanceDiscovery.cs

```diff
@@ -0,0 +1,153 @@
+        public string TenantDiscoveryEndpoint { get; set; }
+
+        [DataMember(Name = "metadata")]
+        public InstanceDiscoveryMetadataEntry[] Metadata { get; set; }
+    }
+
+    [DataContract]
+    internal sealed class InstanceDiscoveryMetadataEntry
+    {
+        [DataMember(Name = "preferred_network")]
+        public string PreferredNetwork { get; set; }
+
+        [DataMember(Name = "preferred_cache")]
+        public string PreferredCache { get; set; }
+
+        [DataMember(Name = "aliases")]
+        public string[] Aliases { get; set; }
+    }
+
+    internal static class InstanceDiscovery
+    {
+        public const string DefaultTrustedAuthority = "login.microsoftonline.com";
+
+        private static HashSet<string> WhitelistedAuthorities = new HashSet<string>(new[]
+        {
+            "login.windows.net", // Microsoft Azure Worldwide - Used in validation scenarios where host is not this list 
+            "login.chinacloudapi.cn", // Microsoft Azure China
+            "login.microsoftonline.de", // Microsoft Azure Blackforest
+            "login-us.microsoftonline.com", // Microsoft Azure US Government - Legacy
+            "login.microsoftonline.us", // Microsoft Azure US Government
+            "login.microsoftonline.com" // Microsoft Azure Worldwide
+        });
+
+        public const string AuthorizeEndpointTemplate = "https://{host}/{tenant}/oauth2/authorize";
+
+        // The following cache could be private, but we keep it public so that internal unit test can take a peek into it
+        public static readonly ConcurrentDictionary<string, InstanceDiscoveryMetadataEntry> InstanceCache =
+            new ConcurrentDictionary<string, InstanceDiscoveryMetadataEntry>(); // Keys are host strings
+
+        private static SemaphoreSlim semaphore = new SemaphoreSlim(1, 1);
+
+        public static async Task<InstanceDiscoveryMetadataEntry> GetMetadataEntry(string host, bool validateAuthority,
+            CallState callState)
+        {
+            InstanceDiscoveryMetadataEntry entry = null;
+            if (!InstanceCache.TryGetValue(host, out entry))
+            {
+                await semaphore.WaitAsync().ConfigureAwait(false); // SemaphoreSlim.WaitAsync() will not block current thread
+                try
+                {
+                    if (!InstanceCache.TryGetValue(host, out entry))
+                    {
+                        await DiscoverAsync(host, validateAuthority, callState).ConfigureAwait(false);
+                        InstanceCache.TryGetValue(host, out entry);
+                    }
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+
+            return entry;
+        }
+
+        private static async Task DiscoverAsync(string host, bool validateAuthority,
+            CallState callState) // No return value. Modifies InstanceCache directly.
```

Comment:
I noticed you are using a commenting style - comments inline.
Other places in the project, the comments are on a separate line.

Reply:
Yeah [I personally prefer inline comment](https://stackoverflow.com/a/13761377/728675), no matter what language I'm using. Because this way the explanation can stick with the very line that I want to comment on; also it would be relatively less likely for the code and comment to become out-of-sync.

In this particular file, there are places that I think inline comments make more sense. Such as [comments on each whitelisted host](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/796/files#diff-90101aa81ac2c8a15a2a9e43b58d79dfR66), and [this comment is actually documenting that very line](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/796/files#diff-90101aa81ac2c8a15a2a9e43b58d79dfR88).

All that being said, I have no problem to modify them in next commit.

Reply:
@rayluo - thanks for following the existing guidelines.